pallets/stateful-storage/src/stateful_child_tree.rs

```diff
@@ -0,0 +1,395 @@
+use core::marker::PhantomData;
+
+use codec::{Decode, Encode};
+use common_primitives::msa::MessageSourceId;
+use frame_support::{
+	storage::{child, child::ChildInfo, ChildTriePrefixIterator},
+	Blake2_128, Blake2_256, Hashable, StorageHasher, Twox128, Twox256,
+};
+use sp_core::{ConstU8, Get};
+use sp_io::hashing::twox_64;
+use sp_std::{fmt::Debug, prelude::*};
+
+pub trait MultipartKeyStorageHasher: StorageHasher {
+	type HashSize: Get<u8>;
+	fn reverse(x: &[u8], num_key_parts: u8) -> &[u8] {
+		let hash_len: usize = (Self::HashSize::get() * num_key_parts).into();
+		if x.len() < hash_len {
+			log::error!("Invalid reverse: hash length too short");
+			return &[]
+		}
+		&x[hash_len..]
+	}
+}
+/// 128-bit Blake2 hash.
+impl MultipartKeyStorageHasher for Blake2_128 {
+	type HashSize = ConstU8<16>;
+}
+/// 256-bit Blake2 hash.
+impl MultipartKeyStorageHasher for Blake2_256 {
+	type HashSize = ConstU8<32>;
+}
+/// 128-bit XX hash.
+impl MultipartKeyStorageHasher for Twox128 {
+	type HashSize = ConstU8<16>;
+}
+/// 256-bit XX hash.
+impl MultipartKeyStorageHasher for Twox256 {
+	type HashSize = ConstU8<32>;
+}
+
+pub trait MultipartStorageKeyPart:
+	Clone + Debug + Default + Decode + Encode + Eq + PartialEq + Hashable
+{
+}
+impl<T> MultipartStorageKeyPart for T where
+	T: Clone + Debug + Default + Decode + Encode + Eq + PartialEq + Hashable
+{
+}
+
+pub trait MultipartKey<H: MultipartKeyStorageHasher>: MultipartStorageKeyPart {
+	type Arity: Get<u8>;
```

Comment:
In Cs arity means:
> in logic, mathematics, and computer science, arity is the number of arguments or operands taken by a function, operation or relation. 

I am trying to draw the connection between the definition and use. Can you please help understand?

Reply:
maybe we can remove this from here and use the `arity` function to implement what it needs.

Reply:
Arity is the length of the tuple as you mentioned. It's the number of keys. Rust allows different kinds of implementations for doing the same thing. What is the advantage of changing this to a function?

Reply:
It's minor; however, notice that when the trait is implemented on a tuple, you also provide the tuple length. The tuple you implemented this trait knows its own length.

Reply:
This is Joe's implementation so I'm not sure why he has chosen this approach but from my understanding you can not get the length of a tuple